src/controllers/PartialUserFormVerifyController.php

```diff
@@ -0,0 +1,114 @@
+<?php
+
+
+namespace Firesphere\PartialUserforms\Controllers;
+
+use Firesphere\PartialUserforms\Forms\PasswordForm;
+use Firesphere\PartialUserforms\Models\PartialFormSubmission;
+use Page;
+use PageController;
+use SilverStripe\Control\HTTPResponse;
+use SilverStripe\Control\HTTPResponse_Exception;
+use SilverStripe\UserForms\Model\UserDefinedForm;
+
+/**
+ * Class \Firesphere\PartialUserforms\Controllers\PartialUserFormVerifyController
+ *
+ */
+class PartialUserFormVerifyController extends PageController
+{
+
+    /**
+     * @var array
+     */
+    private static $allowed_actions = [
+        'getForm'
+    ];
+    /**
+     * @var PartialFormSubmission
+     */
+    protected $partialFormSubmission;
+
+    /**
+     * @var PasswordForm
+     */
+    protected $form;
+
+    /**
+     * @return PartialUserFormVerifyController|void
+     * @throws HTTPResponse_Exception
+     */
+    public function init()
+    {
+        parent::init();
+        $session = $this->getRequest()->getSession();
+        // Set the session if the last session has expired
+        if (!$session->get(PartialSubmissionController::SESSION_KEY)) {
+            return $this->httpError(404);
+        }
+
+        /** @var PartialFormSubmission $partial */
+        $partial = PartialFormSubmission::get()->byID($session->get(PartialSubmissionController::SESSION_KEY));
+
+        $this->setPartialFormSubmission($partial);
+        // Set data record and load the form
+        /** @var UserDefinedForm dataRecord */
+        $this->dataRecord = Page::create();
+    }
+
+    /**
+     * @return PasswordForm
+     */
+    public function getForm()
+    {
+        return PasswordForm::create($this, __FUNCTION__);
+    }
+
+
+    /**
+     * @param array $data
+     * @param PasswordForm $form
+     * @return HTTPResponse
+     */
+    public function doValidate($data, $form)
+    {
+        /** @var PartialFormSubmission $partial */
+        $partial = $this->getPartialFormSubmission();
+
+        $password = hash_pbkdf2('SHA256', $data['Password'], $partial->TokenSalt, 1000);
+        if (!hash_equals($password, $partial->Password)) {
+            $form->sessionError(
+                _t(
+                    PasswordForm::class . '.PASSWORDERROR',
+                    'Password incorrect, please check your password and try again'
+                )
+            );
+
+            return $this->redirectBack();
+        }
+
+        $request = $this->getRequest();
+        $request->getSession()->set(PasswordForm::PASSWORD_SESSION_KEY, $partial->ID);
+
+        $token = $partial->Token;
+        $key = $partial->generateKey($token);
+
+        return $this->redirect(sprintf('/%s/%s/%s', 'partial', $key, $token));
```

Comment:
Can you use `$partial->getPartialLink()`?

Reply:
Yep